include/seqan3/io/alignment_file/format_sam.hpp

```diff
@@ -135,6 +140,280 @@ class alignment_file_format_sam
+                }
+            }
+        }
+        else
+        {
+            std::ranges::next(std::ranges::begin(stream_view)); // skip '*'
+        }
+        std::ranges::next(std::ranges::begin(stream_view));
+
+        // Field 11:  Quality
+        // -------------------------------------------------------------------------------------------------------------
+        auto const tab_or_end = is_char<'\t'> || is_char<'\r'> || is_char<'\n'>;
+        read_field(stream_view | view::take_until_or_throw(tab_or_end), qual);
+
+        // All remaining optional fields if any: SAM tags dictionary
+        // -------------------------------------------------------------------------------------------------------------
+        while (is_char<'\t'>(*std::ranges::begin(stream_view))) // read all tags if present
+        {
+            std::ranges::next(std::ranges::begin(stream_view));
+            read_field(stream_view | view::take_until_or_throw(tab_or_end), tag_dict);
+        }
+
+        if (is_char<'\r'>(*std::ranges::begin(stream_view)))    // protect against '\r\n'
+            std::ranges::next(std::ranges::begin(stream_view)); // skip '\r'
+        std::ranges::next(std::ranges::begin(stream_view));     // skip newline
+
+        // DONE READING - wrap up
+        // -------------------------------------------------------------------------------------------------------------
+        // make sure "buffer at end" implies "stream at end"
+        if (std::ranges::empty(stream_view) && (!stream.eof()))
+            stream.get(); // triggers error in stream and sets eof
+
+        // Alignment object construction
+        // Note that the query sequence in get<1>(align) has already been filled while reading Field 10.
+        if constexpr (!detail::decays_to_ignore_v<align_type>)
+        {
+            if (!cigar.empty() && !std::ranges::empty(get<1>(align))) // only parse alignment if cigar and seq was given
+            {
+                if constexpr (!detail::decays_to_ignore_v<ref_id_to_pos_map_type>)
+                {
+                    static_assert(!detail::decays_to_ignore_v<ref_seqs_type>,
+                                  "If you pass the id_to_ref_map you must also pass ref_seqs.");
+
+                    size_t pos; // get index for ref_seqs
+                    try
+                    {
+                        if constexpr (detail::decays_to_ignore_v<ref_id_type>)
+                            pos = ref_id_to_pos_map.at(ref_id_tmp);
+                        else
+                            pos = ref_id_to_pos_map.at(ref_id);
+                    }
+                    catch (std::out_of_range & e)
+                    {
+                        throw parse_error{std::string("[CORRUPTED SAM FILE] An id in the SAM record cannot be found "
+                                                      "in the given list of reference ids. Check your SAM file!")};
+                    }
```

Comment:
I don't think we have any catch block anywhere in the library, yet, and I think our policy was that we can throw but not rely on catch, right?

Reply:
It depends on the use case. Sometimes you want to catch some exception and rethrow it as a nested exception with some additional information. So I guess this would be the use case here? In this case I am not against it sa it clearly improves the user error. But as I said, better pass it as a nested exception. See here https://en.cppreference.com/w/cpp/error/nested_exception

Reply:
I am pretty sure we agreed on "its ok to throw but not catch inside the library" for performance reasons... We really need to write these things down so that we don't start arguments every time :disappointed:

Reply:
AFAIK, the performance is only affected if an exception is thrown. The argument  is that if an exception is thrown the performance of the program does not matter, as it has to deal with an erroneous state anyway and giving any runtime guarantees in this state do not make a lot of sense. If an exception is not thrown it has zero-costs overhead. So with this regard the catch should not have any impact on the performance unless an exception was thrown

Reply:
I added a card to discuss this.